### ISSUE TITLE
Pkcs9SigningTime has different behaviour for dates <1950 and >2050

### ISSUE THREAD

**Issue by MaximLipnin** [dotnet/corefx#30578]
For some edge cases like
```
DateTime dt = new DateTime(1949, 12, 31, 11, 59, 59);
Pkcs9SigningTime st = new Pkcs9SigningTime(dt);
```
or
```
DateTime dt = new DateTime(2050, 01, 01, 00, 00, 00);
Pkcs9SigningTime st = new Pkcs9SigningTime(dt);
```
it throws an exceptions on Windows and works on MacOS.

**Comment by filipnavara** [comment]
The actual ASN.1 representation of the date uses only two digits for the year, which explains the 1950 to 2049 range. The APIs should behave consistently and throw exceptions for dates outside of the range. Right now it produces values that cannot be round-tripped.

**Comment by filipnavara** [comment]
@karelz I think it's more of a bug than a question :) I stated the reason in the comment above.

**Comment by filipnavara** [comment]
Here's an excerpt from RFC 5652 (Cryptographic Message Syntax):

> Signing-time attribute values have ASN.1 type SigningTime:
> 
>       SigningTime ::= Time
> 
>       Time ::= CHOICE {
>         utcTime UTCTime,
>         generalizedTime GeneralizedTime }
> 
>    Note: The definition of Time matches the one specified in the 1997
>    version of X.509 [X.509-97].
> 
>    Dates between 1 January 1950 and 31 December 2049 (inclusive) MUST be
>    encoded as UTCTime.  Any dates with year values before 1950 or after
>    2049 MUST be encoded as GeneralizedTime.
> 
>    UTCTime values MUST be expressed in Coordinated Universal Time
>    (formerly known as Greenwich Mean Time (GMT) and Zulu clock time) and
>    MUST include seconds (i.e., times are YYMMDDHHMMSSZ), even where the
>    number of seconds is zero.  Midnight MUST be represented as
>    "YYMMDD000000Z".  Century information is implicit, and the century
>    MUST be determined as follows:
> 
>       Where YY is greater than or equal to 50, the year MUST be
>       interpreted as 19YY; and
> 
>       Where YY is less than 50, the year MUST be interpreted as 20YY.
> 
>    GeneralizedTime values MUST be expressed in Coordinated Universal
>    Time and MUST include seconds (i.e., times are YYYYMMDDHHMMSSZ), even
>    where the number of seconds is zero.  GeneralizedTime values MUST NOT
>    include fractional seconds.

Essentially it boils down to two problems:

1. `AsnWriter.WriteUtcTime` should throw exceptions for dates outside of the range.
2. `Pkcs9SigningTime` should support dates outside of the range by using generalized time. However, this is not currently supported on any platform or NetFX.

cc @bartonjs

**Comment by bartonjs** [comment]
> AsnWriter.WriteUtcTime should throw exceptions for dates outside of the range.

No, UTCTime has no particular bounds (in X.680/X.690).  That's a construct of the X.509 Time type.  UTCTime is "given that you know what the century means, write down a two digit year", and the reader just has to agree on what the century means.

An overload could be added which takes the smallest legal year to ensure that it's bounded within a century.

> Pkcs9SigningTime should support dates outside of the range by using generalized time.

Looks like Windows passes `PKCS_UTC_TIME` when it meant `X509_CHOICE_OF_TIME`.  And then the managed implementation should use the X509 Time interpretation as well.

And, of course, tests :smile:.
